<launch>
	<!-- IMU NODE -->
	<node pkg="barc" type="imu_data_acquisition.py" name="imu_node" >
		<param name="port" value="/dev/ttyACM0" />
	</node>

	<!-- ARDUINO NODE -->
	<node pkg="rosserial_python" type="serial_node.py" name="arduino_node" >
		<param name="port" value="/dev/ttyUSB0" />
	</node>

	<!-- STATE ESTIMATION NODE -->
	<node pkg="barc" type="state_estimation.py" name="state_estimation" >
    <!-- VEHICLE PROPERTIES -->
    <!-- m      := mass of vehicle
         I_z    := moment of inertia about z-axis
         L_a    := distance from CoG to rear axel
         L_b    := distance from CoG to front axel -->
		<param name="L_a"   value="0.125" />
		<param name="L_b"   value="0.125" />
		<param name="m"     value="1.98" />
		<param name="I_z"   value="0.24" />

    <!-- EXTERNAL FORCES 
        Ff      := force of friction       -->
        <param name="air_drag_coeff"    value="0.1308"/>
        <param name="Ff"    value="0.1711"/>

    <!-- ENCODER PARAMETERS -->
		<param name="dt_vx" value="0.2" />

    <!-- LOW PASS FILTER PARAMETER -->
        <param name="p_filter" value="0.2" />

    <!-- TIRE MODEL PARAMETERS-->
    <!-- f_paj = mu*m*g/2*sin(c*atan(b*alpha))
        assuming lumped tire model    -->
		<param name="B" value="7.40" />
		<param name="C" value="1.25" />
		<param name="mu" value="0.234" />

    <!-- STATE ESTIMATION PROPERTIES -->
    <!-- Luemberger observer
        aph := estimation error dynamics turning parameter -->
		<param name="aph" value="3.5" />

    <!-- EKF observer
        q   := standard deviation of measurement noise
        r   := standard deviation of process noise
        v_x_min   := minimum velocity before activating EKF -->
		<param name="q" value="0.1" />
		<param name="r" value="0.1" />
		<param name="v_x_min" value="0.3" />
	</node>
	

	<!-- LQR initization MANUEVER -->
	<node pkg="barc" type="controller_hardTurn.py" name="controller" >
            <!--  SELECTION -->
            <param name="user" value="jgon13" />
            <param name="experiment_sel" type="int" value="6" />

            <!-- PROPERTIES
            * v_x_pwm: signal for motor speed
            * steering_angle : signal for servo steering angle
            * t_exp : length of experiment
            * t_turn : time at which motor first turns
            -->
            <param name="v_x_pwm" type="int" value="98" />
            <param name="steering_angle" type="int" value="25" />
            <param name="t_turn" type="int" value="6" />
            <param name="t_exp" type="int" value="7" />
 		
         <!-- PID for straight test using imu gyro -->
            <param name="p" type="double" value="50" />
		    <param name="i" type="double" value="5" />
		    <param name="d" type="double" value="0" />
   </node>

    <!-- Record the experiment data -->
    <node pkg="rosbag" type="record" name="rosbag_record"
        args=" --all" />

</launch>
